### PR TITLE
[ws] Extend VerifyClientCallbackAsync to support code and message fields

### PR DESCRIPTION
**Pull request by akerust**
- [OK] Use a meaningful title for the pull request. Include the name of the package modified.
- [OK] Test the change in your own code. (Compile and run.)
- [OK] Follow the advice from the [readme]
- [OK] Avoid [common mistakes](https://github.com/DefinitelyTyped/DefinitelyTyped/blob/master/README.md#common-mistakes).
- [OK] Run `npm run lint package-name` (or `tsc` if no `tslint.json` is present).

### CODE REVIEW

File: types/ws/index.d.ts [hunk 1]
```diff
@@ -82,7 +82,7 @@ declare namespace WebSocket {
 
     type VerifyClientCallbackSync = (info: { origin: string; secure: boolean; req: http.IncomingMessage }) => boolean;
     type VerifyClientCallbackAsync = (info: { origin: string; secure: boolean; req: http.IncomingMessage }
-        , callback: (res: boolean) => void) => void;
+        , callback: (res: boolean, code?: number, message?: string) => void) => void;
```
Comment (mhegazy) [inline types/ws/index.d.ts]: a callback parameter should not be optional. this should be:
```ts
callback: (res: boolean, code: number, message: string) => void) => void;
```

Comment (mailto1587) [inline types/ws/index.d.ts]: Hi @mhegazy , thanks for suggestion!

But if I set the rest parameters required, declaration is not backward compatible. All the existing codes cannot pass the compilation.

Comment (mhegazy) [inline types/ws/index.d.ts]: you have an example?

Comment (mailto1587) [inline types/ws/index.d.ts]:
```typescript
const wss = new WebSocket.Server({
    host: address,
    port: port,
    verifyClient: (info, verify) => {
        const id = getIdFromRequest(info.req);  // Get "id" param from URL parameters.
        if (id) {
            verify(true);
        } else {
            verify(false);
        }
    }
});
```

Suppose we have existing code like above, it works with current declaration.

But if we extend the declaration without optional, just like:

```typescript
callback: (res: boolean, code: number, message: string) => void) => void;
```

The tsc compiler will reject the code with error `[ts] Supplied parameters do not match any signature of call target.`

**Comment by aozgaa** [comment]
There are some checklist items you removed from the template. In particular, could you link to to the documentation and provide a test that exhibits the effect of the change?

Based on your example, it looks like the optional parameters may be appropriate. Does the typing mean that someone implementing  `VerifyClientCallbackAsync` may call the callback might be called with just the `res` parameter (cf: http://www.typescriptlang.org/docs/handbook/declaration-files/do-s-and-don-ts.html#optional-parameters-in-callbacks)? Again, documentation would help here immensely.

**Comment by andy-ms** [comment]
I take it that the user may provide one callback argument, but may also provide two or three. Looks good.

**Pull request merged**